베이스라인 측정 파이썬 구현 #1

ene5135 opened the pull request:
울산 코퍼스의 베이스라인을 측정하기 위한 파이썬 프로그램을
작성하였습니다.

코퍼스를 제가 처리하기 쉽도록 전처리 해주는 transform.py와
전처리한 코퍼스를 통해 베이스라인 정확도를 구해주는 baseline.py를
구현하였습니다.

베이스라인은 동형이의어 의미들 중 가장 자주 등장한 의미를 고르는
방식입니다.

베이스라인을 구하기 위해 트레이닝 셋에 나타나는 동형이의어들로 이루어진
어휘사전을 구축(voca.bin)하고 각 동형이의어별로 가장 많이 등장한 의미들만
추려내었습니다(max_freq_dic.bin).

테스트셋의 동형이의어 단어에 대해 답을할때는 가장 자주 등장한 의미를
선택하되 만약 학습한적이 없는 동형이의어라면 UNKNOWN 으로 답해 평가시
틀린것으로 계산되도록 하였습니다.

구체적인 시현방법은 README에 적어두었습니다.

Commit:
This is first commit to test git flow #1

Commit:
베이스라인 측정 파이썬 구현 #1

울산 코퍼스의 베이스라인을 측정하기 위한 파이썬 프로그램을
작성하였습니다.

코퍼스를 제가 처리하기 쉽도록 전처리 해주는 transform.py와
전처리한 코퍼스를 통해 베이스라인 정확도를 구해주는 baseline.py를
구현하였습니다.

베이스라인은 동형이의어 의미들 중 가장 자주 등장한 의미를 고르는
방식입니다.

베이스라인을 구하기 위해 트레이닝 셋에 나타나는 동형이의어들로 이루어진
어휘사전을 구축(voca.bin)하고 각 동형이의어별로 가장 많이 등장한 의미들만
추려내었습니다(max_freq_dic.bin).

테스트셋의 동형이의어 단어에 대해 답을할때는 가장 자주 등장한 의미를
선택하되 만약 학습한적이 없는 동형이의어라면 UNKNOWN 으로 답해 평가시
틀린것으로 계산되도록 하였습니다.

구체적인 시현방법은 README에 적어두도록 하겠습니다.

Commit:
베이스라인 측정 방법 README 추가 #1

readme에 베이스라인 측정방법을 적었습니다.

ene5135 committed:
README 를 조금 수정하였습니다. #1

쉘 입력 예시 코드블럭이 보이도록 수정하였습니다.

krikit commented:
우선 python 스크립트는 커밋하기 전에 [pylint](https://www.pylint.org/)를 돌려서 에러를 잡았으면 해요. 까먹을 수도 있으니 [git-pylint-commit-hook](http://git-pylint-commit-hook.readthedocs.io/en/latest/)을 설치하고, 아래와 같이 .git/hooks/pre-commit 훅 파일에 추가해 주면 좋아요. (pre-commit 파일에 실행 권한을 줘야 합니다.)

```bash
#!/usr/bin/env bash

git-pylint-commit-hook
```

아래는 transform.py 파일의 pylint 결과입니다.

```
No config file found, using default configuration
************* Module transform
C: 20, 0: Exactly one space required after comma
        line = line.replace("\n","")
                                ^ (bad-whitespace)
C: 21, 0: No space allowed before :
        if REMOVE_SENSE : line = re.sub('__[\d][\d]',"",line)
                        ^ (bad-whitespace)
C: 21, 0: Exactly one space required after comma
        if REMOVE_SENSE : line = re.sub('__[\d][\d]',"",line)
                                                    ^ (bad-whitespace)
C: 21, 0: Exactly one space required after comma
        if REMOVE_SENSE : line = re.sub('__[\d][\d]',"",line)
                                                       ^ (bad-whitespace)
C: 22, 0: Exactly one space required after comma
        line = re.split('[+ ]',line)
                              ^ (bad-whitespace)
C: 28, 0: Trailing newlines (trailing-newlines)
W: 21, 0: Anomalous backslash in string: '\d'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
W: 21, 0: Anomalous backslash in string: '\d'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
C:  1, 0: Missing module docstring (missing-docstring)
W: 11, 0: No exception type(s) specified (bare-except)
C:  9, 4: Invalid constant name "fr" (invalid-name)
C: 10, 4: Invalid constant name "fw" (invalid-name)
C: 15, 0: Invalid constant name "count" (invalid-name)
C: 17, 4: Invalid constant name "line" (invalid-name)
C: 18,17: More than one statement on a single line (multiple-statements)
C: 20, 8: Invalid constant name "line" (invalid-name)
C: 21,26: More than one statement on a single line (multiple-statements)
C: 21,26: Invalid constant name "line" (invalid-name)
C: 22, 8: Invalid constant name "line" (invalid-name)
C: 24, 4: Invalid constant name "count" (invalid-name)

-----------------------------------
Your code has been rated at 2.00/10
```

그리고 아래는 baseline.py 파일의 pylint 결과입니다.

```
No config file found, using default configuration
************* Module baseline
C:  9, 0: Trailing whitespace (trailing-whitespace)
C: 41, 0: Exactly one space required after comma
        line = line.replace("\n","")
                                ^ (bad-whitespace)
C: 42, 0: Exactly one space required after comma
        newTokens = re.split('[ ]',line)
                                  ^ (bad-whitespace)
C: 46, 0: Trailing whitespace (trailing-whitespace)
C: 58, 0: Line too long (122/100) (line-too-long)
C: 65, 0: Trailing whitespace (trailing-whitespace)
C: 71, 0: Exactly one space required after comma
        splitKey = re.split('/',key)
                               ^ (bad-whitespace)
C: 77, 0: Exactly one space required after comma
        for n in range(1,99):
                        ^ (bad-whitespace)
C: 78, 0: No space allowed before :
            if n < 10 :
                      ^ (bad-whitespace)
C: 80, 0: No space allowed before :
            else :
                 ^ (bad-whitespace)
C: 87, 0: Line too long (101/100) (line-too-long)
C: 92, 0: Trailing whitespace (trailing-whitespace)
C: 94, 0: Trailing whitespace (trailing-whitespace)
C:102, 0: Line too long (147/100) (line-too-long)
C:114, 0: Exactly one space required after comma
        line = line.replace("\n","")
                                ^ (bad-whitespace)
C:115, 0: Exactly one space required after comma
        tokens = re.split('[ ]',line)
                               ^ (bad-whitespace)
C:143, 0: Exactly one space required after comma
        testLine = fr_test.readline().replace("\n","")
                                                  ^ (bad-whitespace)
C:144, 0: Exactly one space required after comma
        answerLine = fr_answer.readline().replace("\n","")
                                                      ^ (bad-whitespace)
C:146, 0: No space allowed before :
        if not testLine or not answerLine : break
                                          ^ (bad-whitespace)
C:159, 0: Exactly one space required after comma
    print("The number of homograph : ",count)
                                      ^ (bad-whitespace)
C:160, 0: Exactly one space required after comma
    print("The number of correct answer : ",correct)
                                           ^ (bad-whitespace)
C:161, 0: Exactly one space required after comma
    print("Precision : ",(correct/count)*100)
                        ^ (bad-whitespace)
C:162, 0: Trailing whitespace (trailing-whitespace)
C:177, 0: Trailing newlines (trailing-newlines)
W: 44, 0: Anomalous backslash in string: '\d'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
W: 44, 0: Anomalous backslash in string: '\d'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
W: 68, 0: Anomalous backslash in string: '\d'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
W: 68, 0: Anomalous backslash in string: '\d'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
W:118, 0: Anomalous backslash in string: '\d'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
W:118, 0: Anomalous backslash in string: '\d'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
W:119, 0: Anomalous backslash in string: '\d'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
W:119, 0: Anomalous backslash in string: '\d'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
W:154, 0: Anomalous backslash in string: '\d'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
W:154, 0: Anomalous backslash in string: '\d'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
C:  1, 0: Missing module docstring (missing-docstring)
E:  2, 0: Unable to import 'nltk.tokenize' (import-error)
E:  3, 0: Unable to import 'nltk' (import-error)
W:  8, 0: String statement has no effect (pointless-string-statement)
W: 12, 0: String statement has no effect (pointless-string-statement)
W: 13, 0: String statement has no effect (pointless-string-statement)
W: 23, 0: No exception type(s) specified (bare-except)
W: 30, 0: String statement has no effect (pointless-string-statement)
W: 31, 0: String statement has no effect (pointless-string-statement)
W: 32, 0: String statement has no effect (pointless-string-statement)
C: 34, 0: Invalid function name "buildVoca" (invalid-name)
C: 34, 0: Missing function docstring (missing-docstring)
C: 35, 4: Invalid variable name "fr" (invalid-name)
C: 37, 4: Invalid variable name "tokenList" (invalid-name)
C: 40,21: More than one statement on a single line (multiple-statements)
C: 42, 8: Invalid variable name "newTokens" (invalid-name)
W: 57, 0: String statement has no effect (pointless-string-statement)
W: 58, 0: String statement has no effect (pointless-string-statement)
W: 59, 0: String statement has no effect (pointless-string-statement)
W: 75, 8: Redefining built-in 'max' (redefined-builtin)
C: 61, 0: Invalid function name "buildMaxFreqDic" (invalid-name)
C: 61, 0: Missing function docstring (missing-docstring)
C: 63, 4: Invalid variable name "fw_maxFreqDic" (invalid-name)
C: 68, 4: Invalid variable name "keysWithoutSense" (invalid-name)
C: 69, 4: Invalid variable name "maxFreqDic" (invalid-name)
C: 71, 8: Invalid variable name "splitKey" (invalid-name)
C: 76, 8: Invalid variable name "maxSense" (invalid-name)
C: 77,12: Invalid variable name "n" (invalid-name)
C: 79,16: Invalid variable name "targetWord" (invalid-name)
C: 81,16: Invalid variable name "targetWord" (invalid-name)
C: 85,16: Invalid variable name "maxSense" (invalid-name)
W:100, 0: String statement has no effect (pointless-string-statement)
W:101, 0: String statement has no effect (pointless-string-statement)
W:102, 0: String statement has no effect (pointless-string-statement)
W:103, 0: String statement has no effect (pointless-string-statement)
C:105, 0: Invalid function name "makeAnswer" (invalid-name)
C:105, 0: Missing function docstring (missing-docstring)
C:108, 4: Invalid variable name "fr_maxFreqDic" (invalid-name)
C:109, 4: Invalid variable name "maxFreqDic" (invalid-name)
C:113,21: More than one statement on a single line (multiple-statements)
C:119,16: Invalid variable name "queryWord" (invalid-name)
C:120,16: Invalid variable name "answerWord" (invalid-name)
C:122,16: Invalid variable name "answerWord" (invalid-name)
W:133, 0: String statement has no effect (pointless-string-statement)
W:134, 0: String statement has no effect (pointless-string-statement)
C:135, 0: Missing function docstring (missing-docstring)
C:143, 8: Invalid variable name "testLine" (invalid-name)
C:144, 8: Invalid variable name "answerLine" (invalid-name)
C:146,44: More than one statement on a single line (multiple-statements)
C:148, 8: Invalid variable name "testTokens" (invalid-name)
C:149, 8: Invalid variable name "answerTokens" (invalid-name)
C:151, 8: Invalid variable name "mergedTokens" (invalid-name)
W:  2, 0: Unused word_tokenize imported from nltk.tokenize (unused-import)
C:  4, 0: standard import "import pickle" should be placed before "from nltk.tokenize import word_tokenize" (wrong-import-order)
C:  5, 0: standard import "import os" should be placed before "from nltk.tokenize import word_tokenize" (wrong-import-order)
C:  6, 0: standard import "import sys" should be placed before "from nltk.tokenize import word_tokenize" (wrong-import-order)

-----------------------------------
Your code has been rated at 2.16/10
```

krikit commented:
참고할만한 Python 3 코드 템플릿을 하나 공유 드릴게요. 아래 코드는 [PEP8](https://www.python.org/dev/peps/pep-0008)에 맞춰져 있고, pylint safe 합니다. ^^;

https://gist.github.com/krikit/9aaa60d24d977f92cefdafdd425dd523

krikit commented on baseline/baseline.py:
```diff
@@ -0,0 +1,177 @@
+import re
+from nltk.tokenize import word_tokenize
+import nltk
```
nltk 패키지를 사용했네요. 이렇게 다른 패키지에 의존성이 생긴 경우 해당 패키지를 설치하도록 의존 패키지를 나열해 주는 것이 좋습니다.

보통 git 저장소의 루트에 requirements.txt 파일을 두고 아래와 같이 의존 패키지를 나열합니다.

```
nltk
numpy
pandas
tensorflow
...
```

이렇게 하면 git 저장소를 내려받은 사람은 requirements.txt 파일을 이용해 의존 패키지를 한꺼번에 설치할 수 있습니다.

```bash
# using pip
pip install -r requirements.txt

# using conda
conda install --yes --file requirements.txt
```

krikit commented on baseline/baseline.py:
```diff
@@ -0,0 +1,177 @@
+''' train_set.txt and test_set.txt should be in current directory '''
+
+THIS_FOLDER = os.path.dirname(os.path.abspath(__file__))
+VOCA = os.path.join(THIS_FOLDER, 'voca.bin')
```
voca.bin과 max_freq_dic.bin 파일은 코퍼스를 통해 생성할 수 있는 2차 생성물이고, 바이너리 파일이므로 굳이 git 저장소에 추가하지 않아도 좋을 듯 합니다.

krikit commented on baseline/baseline.py:
```diff
@@ -0,0 +1,177 @@
+
+    print("The number of homograph : ",count)
+    print("The number of correct answer : ",correct)
+    print("Precision : ",(correct/count)*100)
```
[precision은 recall](https://en.wikipedia.org/wiki/Precision_and_recall)과 함께 쓰이는 용어로 여기서는 [accuracy](https://en.wikipedia.org/wiki/Accuracy_and_precision)가 맞습니다.

Commit:
코드 스타일 수정하고 .gitignore 와 requirement 추가하였습니다. #1

코드 스타일 pylint pep8 스타일에 맞추어 수정하였습니다.
requirement.txt 에 nltk 추가해두었습니다.
gitignore에 bin 파일 무시하도록 하였습니다.
precision 용어를 accuracy로 수정하였습니다.

Commit:
한글자 짜리 변수 'n' 을 'num' 으로 수정하였습니다. #1

krikit commented on baseline/baseline.py:
```diff
@@ -0,0 +1,195 @@
+    fr_test.close()
+    fr_answer.close()
+
+build_voca()
```
여기서부터는 main() 함수를 만들어 그 아래에 넣는 것이 좋을 것 같아요.
그리고 제일 마지막에 아래와 같이 main() 함수를 호출하는 식으로요.

```python
def main():
    ...

if __name__ == '__main__':
    main()
```

시행 가능한 python 스크립트를 작성할 때 흔히들 사용하는 python의 idiom 입니다. ^^;

krikit commented on baseline/transform.py:
```diff
@@ -0,0 +1,34 @@
+THIS_FOLDER = os.path.dirname(os.path.abspath(__file__))
+REMOVE_SENSE = False
+
+try:
```
여기서부터도 main() 함수 안에 넣는 것이 좋을 것 같아요. 모든 변수들이 top level에 있어서 pylint는 global 변수를 상수에만 허용해서 대문자를 요구한 거거든요.

ene5135 replied:
아하.. 제가 파이썬이 익숙치않아서 부족한게 많네요 ㅠ 수정하도록 하겠습니다

Commit:
main 함수를 추가하였습니다. #1

krikit commented on baseline/transform.py:
```diff
@@ -0,0 +1,39 @@
+        print('usage: python3 transform.py source_file_name dist_file_name')
+        sys.exit(1)
+
+    count = 0
```
아래 코드는 비효율적이라거나 굳이 수정할 필요는 없지만, 아래와 같이 권장 드리니 다음에 코드를 작성할실 때 참고하세요.

1) writer 파일은 `with` 문을 사용하면 `close()`를 명시적으로 하지 않아도 되요.
2) reader 파일은 iterator라서 `for`문에 바로 쓸 수 있어요.
3) 파일의 `.write()` 메소드 보다 `print()` 함수의 named argument로 `file=fw_dist`와 같이 주는 것이 new line을 별도로 출력하는 수고를 덜 수 있어요.
4) `enumerate`는 리스트 혹은 iterator의 element와 함께 인덱스를 튜플로 만들어주는 유용한 함수이니 기억하면 쓰기 편할 거에요.

```python
    with open(sys.argv[2], 'w') as fw_dist:
        for count, line in enumerate(fr_source):
            line = line.rstrip('\r\n')
            if count % 3 == 1:
                line = line.replace("\n", "")
                if remove_sense:
                    line = re.sub(r'__[\d][\d]', "", line)
                line = re.split('[+ ]', line)
                print(' '.join(line), file=fw_dist)
 ```

krikit commented on baseline/baseline.py:
```diff
@@ -0,0 +1,203 @@
+        max_freq = -1
+        max_sense = "UNKNOWN"
+        for num in range(1, 99):
+            if num < 10:
```
아래와 같이 포맷 함수를 쓰면 되요. ^^

```python
            target_word = '{}__{:02d}/{}'.format(word, num, pos)
```

krikit commented on baseline/baseline.py:
```diff
@@ -0,0 +1,203 @@
+        tokens = re.split('[ ]', line)
+
+        for token in tokens:
+            if re.compile(r'.*__[\d][\d].*').match(token):
```
정규식을 매번 컴파일하지 않고 한번만 컴파일을 하는 것이 속도에 유리하지 않을까요?

```python
TKN_PTN = re.compile(r'.*__[\d][\d].*')    # 상수로 한번만 미리 컴파일

            if PTN.match(token):    # 매번 루프에서는 그냥 매칭만 호출
```

Commit:
꿀팁들을 적용하였습니다. #1

hubert-lee commented:
corpus
배/NNG_01
배/NNG_01
배/NNG_02
배/NNG_02
배/NNG_03
배/NNG_04
배/NNG_04
배/NNG_04
배/NNG_04
배/NNG_05
배/NNG_06
배/NNG_06
배/NNG_06
배/NNG_06
배/NNG_06

vocabulary
배/NNG_01  2
배/NNG_02  2
배/NNG_03  1
배/NNG_04  4
배/NNG_05  1
배/NNG_06  5

max_freq_dict
배/NNG - 배/NNG_06

이렇게 vocabulary를 만들어두고, 이를 이용하여
"배/NNG"에 대해서 _00 ~ _99까지 붙여서 키를 100개를 생성하여 100번을 
탐색한 후, 최대인 "배/NNG_06"을 찾아서
위에 max_freq_dict를 생성합니다.

찾아야할 단어가 100(keys_without_sense의 원소 수)개라면, 100*100의 탐색 횟수가 발생하게 됩니다.
인위적으로 키를 생성해서 불필요하게 비교를 해보는 것보다, 아예 처음부터 코퍼스를 읽어들이면서
아래와 같은 사전구조로 빈도수를 카운트 해두면, 위처럼 불필요한 탐색은 안해도 되지 않을까요?

max_freq_dict["배/NNG"] = {"01":2, "02":2, "03":1, "04":4, "05":1, "06":5}

dictionary에서 key나 value를 대상으로 max를 취하는 방법은 함수로 다 마련되어있을테니
크게 어렵진 않을것 같네요~

ene5135 commented:
감사합니다. 지금보니 제가 너무 생각없이 코드를 짰군요 ㅠ 
일단 해보고 너무 오래걸리면 고쳐야지 하고 놔두었던 부분인데 다음부터는 애초에 잘 생각해서 짜도록 하겠습니다!

말씀해주신대로 처음 코퍼스 읽을때 사전속에 사전식으로 vocabulary를 구축하고 max_freq를 찾으니까 훨씬 빠르게 찾아지는것 같습니다.

```python

def build_voca_with_freq_dic():
    """
    voca : key = word_without_sense, value = freq_dic
    freq_dic : key = sense_number, value = appearing count
    """
    vocabulary = {}
    with open(TRAIN_SET, 'r') as fr_train, open(VOCA, 'wb') as fw_voca:
        for line in fr_train:
            line = line.replace("\n", "")
            new_tokens = re.split('[ ]', line)
            for token in new_tokens:
                if TKN_PTN.match(token):
                    key = re.sub(r'__[\d][\d]', '', token)
                    sense_number = token[token.index("/")-2:token.index("/")]
                    freq_dic = vocabulary.get(key, {})
                    freq_dic[sense_number] = freq_dic.get(sense_number, 0) + 1
                    vocabulary[key] = freq_dic
        pickle.dump(vocabulary, fw_voca)


def build_max_freq_dic2():
    """
    another version of build_max_freq_dic function
    It uses build_voca_with_freq_dic()
    """
    with open(VOCA, 'rb') as fr_voca, open(MAX_FREQ_DIC, 'wb') as fw_max_freq_dic:
        vocabulary = pickle.load(fr_voca)
        max_freq_dic = {}
        for key, freq_dic in vocabulary.items():
            max_freq_dic[key] \
            = key[:key.index("/")] + "__" + max(freq_dic, key=freq_dic.get) + key[key.index("/"):]

        pickle.dump(max_freq_dic, fw_max_freq_dic)


```
